doc/dev/adr/0023-property-support.md

```diff
@@ -0,0 +1,312 @@
+
+public class PropertyListener<TProp> : TelemetryReceiver<TProp>;
+```
+
+The service-side `PropertyMaintainer` is implemented as follows:
+
+```csharp
+public class PropertyWriteResponder<TProp, TStat> : CommandExecutor<TProp, TStat>;
+
+public class PropertyReadResponder<TCtrl, TProp> : CommandExecutor<TCtrl, TProp>;
+
+public class PropertyObserveResponder<TCtrl, TStat> : CommandExecutor<TCtrl, TStat>;
+
+public class PropertyUnobserveResponder<TCtrl, TStat> : CommandExecutor<TCtrl, TStat>;
+
+public class PropertyNotifier<TProp> : TelemetrySender<TProp>;
+```
+
+The *write* action is performed by issuing a 'write' Command request.
+
+* The request payload is an instance of `TProp`.
+  * For statically itemized properties, the optional fields in the `TProp` request object contain values for any Properties that are to be written.
+  * For dynamically itemized properties, the keys in the `TProp` request map indicate which Properties to write, and the corresponding values are what is to be written.
+* The response payload is an instance of `TStat`.
+  * For statically itemized properties, each optional field in the `TStat` response object indicates the status of the write operation for the Property named by the field.
+  * For dynamically itemized properties, each value in the `TStat` response map indicates the status of the write operation for the Property named by the key.
+  * The semantics of `TStatVal` values are determined by the service that implements the property.
+
+The *read* action is performed by issuing a 'read' Command request.
+
+* The request payload is an instance of `TCtrl`.
+  * For statically itemized properties, the optional fields in the `TCtrl` request object are `true` for each field whose Property is to be read.
+  * For dynamically itemized properties, the elements in the `TCtrl` request array indicate the names of the Properties to read.
+* The response payload is an instance of `TProp`.
+  * For statically itemized properties, the optional fields in the `TProp` response object contain values for any Properties that have been read.
+  * For dynamically itemized properties, the keys in the `TProp` response map indicate which Properties have been read, and the corresponding values are what was read.
+
+The *observe* action is performed by issuing an 'observe' Command request.
+
+* The request payload is an instance of `TCtrl`.
+  * For statically itemized properties, the optional fields in the `TCtrl` request object are `true` for each field whose Property is to be added to the notify list.
+  * For dynamically itemized properties, the elements in the `TCtrl` request array indicate the names of the Properties to add to the notify list.
+  * The maintainer may keep a single notify list across all consumers, or it may keep a separate notify list for each consumer; in the latter case, the maintainer must be provided with a way to identify the consumer that requested the observation.
```

Comment:
I would propose that this is functionality that we provide in our SDK (keeping the list of where notifications should be sent), and then having the customer component just deciding whether to use their client id as the scope for setting/receiving notifications or some other shared key as the scope. I still feel that supporting non-consumer specific notifications is a very concerning/confusing scenario, since another consumer can unobserve notifications that you wanted, but I understand this is a scenario we need to support

Reply:
> I would propose that this is functionality that we provide in our SDK

I don't disagree, but I think it is a wholly separable component, maybe a helper class/module that can be used by services that want this functionality.  I don't know if we want to bundle it into our current dev schedule, though.  This is not my call, of course.

> I still feel that supporting non-consumer specific notifications is a very concerning/confusing scenario

It may be, but it is in fact the semantics of OPC UA, so we cannot avoid supporting it (as you noted).